### PR TITLE
Shared metadata index

### PR DESCRIPTION
**Pull request by pauldix**
This PR pulls the in memory index out into something that can be shared across shards. It keeps the information for field encoding local to each shard.

### CODE REVIEW

File: tsdb/shard.go [hunk 1]
```diff
@@ -177,7 +113,8 @@ func (s *Shard) WritePoints(points []Point) error {
 		// marshal the raw data if it hasn't been marshaled already
 		if p.Data() == nil {
 			// this was populated earlier, don't need to validate that it's there.
-			data, err := s.series[p.Key()].measurement.fieldCodec.EncodeFields(p.Fields())
+			warn(p.Name(), s.measurementFields)
```
Comment (benbjohnson) [inline tsdb/shard.go]: Did you mean to include this?

**Comment by benbjohnson** [comment]
I took a cursory look and it lgtm.

File: tsdb/shard.go (hunk 1 above)
Comment (pauldix) [inline tsdb/shard.go]: Bah, this is what happens when you rush to commit right before running out for a fun Saturday evening ;)

File: tsdb/meta.go [hunk 2]
```diff
@@ -16,13 +15,203 @@ const (
 	maxStringLength = 64 * 1024
 )
 
+// Database is the in memory index of a collection of measurements, time series, and their tags.
+type Database struct {
+	// in memory metadata index, built on load and updated when new series come in
```
Comment (dgnorton) [inline tsdb/meta.go]: Should it be called `MetaIndex`?  The comment above and below the `type` describe it that way.

**Comment by dgnorton** [comment]
Is all or most of the code going to move into `package tsdb`?

File: tsdb/meta.go (hunk 2 above)
Comment (pauldix) [inline tsdb/meta.go]: Maybe `DatabaseIndex`? It maps to what a user calls a `Database` in InfluxDB

**Comment by pauldix** [comment]
@dgnorton all of the query processing logic, `database.go`, `shard.go`, `engine.go`, `functions.go`, and `tx.go` will probably be getting moved there.

**Pull request merged**